spec.html

```diff
@@ -37444,29 +37408,53 @@ <h1>Array.prototype.findLast ( _predicate_ [ , _thisArg_ ] )</h1>
-          <p>If a _thisArg_ parameter is provided, it will be used as the *this* value for each invocation of _predicate_. If it is not provided, *undefined* is used instead.</p>
-          <p>_predicate_ is called with three arguments: the element, the index of the element in the array, and the object being traversed.</p>
-          <p>`findLastIndex` does not directly mutate the object on which it is called but the object may be mutated by the calls to _predicate_.</p>
-          <p>The range of elements processed by `findLastIndex` is set before the first call to _predicate_. Elements that are appended to the array after the call to `findLastIndex` begins will not be visited by _predicate_. If existing elements of the array are changed, their value as passed to _predicate_ will be the value at the time that `findLastIndex` visits them.</p>
+          <p>`findLastIndex` calls _predicate_ once for each element of the array, in descending index order, until it finds one where _predicate_ returns *true*. If such an element is found, `findLastIndex` immediately returns the index of that element value. Otherwise, `findLastIndex` returns -1.</p>
+          <p>See ArrayFindViaPredicate for additional information.</p>
         </emu-note>
         <p>This method performs the following steps when called:</p>
         <emu-alg>
-          1. Let _O_ be ? ToObject(*this* value).
-          1. Let _len_ be ? LengthOfArrayLike(_O_).
-          1. If IsCallable(_predicate_) is *false*, throw a *TypeError* exception.
-          1. Let _k_ be _len_ - 1.
-          1. Repeat, while _k_ &ge; 0,
-            1. Let _Pk_ be ! ToString(𝔽(_k_)).
-            1. Let _kValue_ be ? Get(_O_, _Pk_).
-            1. Let _testResult_ be ToBoolean(? Call(_predicate_, _thisArg_, &laquo; _kValue_, 𝔽(_k_), _O_ &raquo;)).
-            1. If _testResult_ is *true*, return 𝔽(_k_).
-            1. Set _k_ to _k_ - 1.
-          1. Return *-1*<sub>𝔽</sub>.
+          1. Let _findRec_ be ? ArrayFindViaPredicate(*this* value, ~descending~, _predicate_, _thisArg_).
+          1. Return _findRec_.[[Index]].
         </emu-alg>
         <emu-note>
           <p>This method is intentionally generic; it does not require that its *this* value be an Array object. Therefore it can be transferred to other kinds of objects for use as a method.</p>
         </emu-note>
+
+        <emu-clause id="sec-arrayfindviapredicate" type="abstract operation">
+          <h1>
+            ArrayFindViaPredicate (
+              _V_: an ECMAScript language value,
+              _direction_: ~ascending~ or ~descending~,
+              _predicate_: an ECMAScript language value,
```

Comment:
seems kind of silly to not restrict the AO to taking a function - it seems like it'd be cleaner to have all the callers check for this.

Reply:
Yup, could do, but that's step 3, and both steps 1 and 2 could throw, so the callers would have to handle all 3 steps in order to preserve the priority of throws. (They'd be handling steps 1 and 2 anyway *if* we go for FindViaPredicate with bullet 2.)

Reply:
ah, good point - probably best to ignore this comment pending the editors decision on the combined predicate.